File: app/src/main/java/ir/apptune/calendar/widget/AppWidget.java
package ir.apptune.calendar.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;

import ir.apptune.calendar.CalendarTool;
import ir.apptune.calendar.EnglishMonthName;
import ir.apptune.calendar.MainActivity;
import ir.apptune.calendar.PersianMonthName;
import ir.apptune.calendar.PersianNumberFormatHelper;
import ir.apptune.calendar.R;

/**
 * The CLass that widget uses to show data
 */

public class AppWidget extends AppWidgetProvider {


    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        super.onUpdate(context, appWidgetManager, appWidgetIds);


        for (int i = 0; i < appWidgetIds.length; i++) {
            int appWidgetId = appWidgetIds[i];
            Intent intent = new Intent(context, MainActivity.class);
            PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, 0);
            RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.widget_layout);
            remoteViews.setOnClickPendingIntent(R.id.layout_widget, pendingIntent);
            CalendarTool calendar = new CalendarTool();
            String ps = calendar.getWeekDayStr() + " " + PersianNumberFormatHelper.toPersianNumber(calendar.getIranianDay() + "") + " " +
                    PersianMonthName.getName(calendar.getIranianMonth()) + " " +
                    PersianNumberFormatHelper.toPersianNumber(calendar.getIranianYear() + "");
            String ms = PersianNumberFormatHelper.toPersianNumber(calendar.getGregorianDay() + "") + " " + EnglishMonthName.getName(calendar.getGregorianMonth()) + " " +
                    PersianNumberFormatHelper.toPersianNumber(calendar.getGregorianYear() + "");
            remoteViews.setTextViewText(R.id.txt_widget_show_persian_date, ps);
            remoteViews.setTextViewText(R.id.txt_widget_show_miladi_date, ms);
            appWidgetManager.updateAppWidget(appWidgetId, remoteViews);


        }
    }
}

File: app/src/main/java/ir/apptune/calendar/features/widget/AppWidget.kt
package ir.apptune.calendar.features.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import dagger.hilt.android.AndroidEntryPoint
import ir.apptune.calendar.MainActivity
import ir.apptune.calendar.R
import ir.apptune.calendar.pojo.CalendarModel
import ir.apptune.calendar.utils.extensions.toEnglishMonth
import ir.apptune.calendar.utils.extensions.toPersianMonth
import ir.apptune.calendar.utils.extensions.toPersianNumber
import ir.apptune.calendar.utils.extensions.toPersianWeekDay
import javax.inject.Inject

/**
 * The CLass that widget uses to show data
 */
@AndroidEntryPoint
class AppWidget : AppWidgetProvider() {

    @Inject
    lateinit var today: CalendarModel

    override fun onUpdate(context: Context, appWidgetManager: AppWidgetManager, appWidgetIds: IntArray) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)

        val intent = Intent(context, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(context, 0, intent, 0)
        val remoteViews = RemoteViews(context.packageName, R.layout.widget_layout).apply {
            setOnClickPendingIntent(R.id.layout_widget, pendingIntent)
        }

        with(today) {
            remoteViews.setTextViewText(R.id.txtWidgetPersianDate, context.getString(R.string.persian_full_date,
                    dayOfWeek.toPersianWeekDay(context),
                    iranianDay.toPersianNumber(),
                    iranianMonth.toPersianMonth(context),
                    iranianYear.toPersianNumber()))
            remoteViews.setTextViewText(R.id.txtWidgetGregorianDate, context.getString(R.string.gregorian_full_date,
                    gDay.toPersianNumber(),
                    gMonth.toEnglishMonth(context),
                    gYear.toPersianNumber()))
        }

        appWidgetIds.forEach { appWidgetManager.updateAppWidget(it, remoteViews) }
    }
}